added asynchronous learning tutorial

joeljosephjin opened the pull request:
As part of increasing the documentation, i have added explanation of some differences in the code wrt to the q-learning example. And then i provided a full sample code for running the algorithm right away. Hope its useful :)

favre49 commented on doc/tutorials/reinforcement_learning/reinforcement_learning.txt:
```diff
@@ -231,9 +233,172 @@ to have converged when the average return reaches a predetermined value (i.e. >
+agents are referred to as "workers". Currently mlpack has One-Step Q-Learning worker, N-Step 
+Q-Learning worker and One-Step SARSA worker.
+
+Lets examine the sample code in chunks.
```
```suggestion
Let's examine the sample code in chunks.
```

favre49 commented on doc/tutorials/reinforcement_learning/reinforcement_learning.txt:
```diff
@@ -231,9 +233,172 @@ to have converged when the average return reaches a predetermined value (i.e. >
+#include <mlpack/methods/reinforcement_learning/policy/aggregated_policy.hpp>
+@endcode
+
+Here we don't use experience replay. And instead of a single policy, we use three different
```
```suggestion
Here we don't use experience replay, and instead of a single policy, we use three different
```

favre49 commented on doc/tutorials/reinforcement_learning/reinforcement_learning.txt:
```diff
@@ -231,9 +233,172 @@ to have converged when the average return reaches a predetermined value (i.e. >
+
+@code
+OneStepQLearning<
+    CartPole, decltype(model), ens::AdamUpdate, decltype(policy)>
```
I believe we can fit this in the line above

favre49 commented on doc/tutorials/reinforcement_learning/reinforcement_learning.txt:
```diff
@@ -231,9 +233,172 @@ to have converged when the average return reaches a predetermined value (i.e. >
+Learning class inside a for loop which runs for 100 training episodes.
+
+@code
+for(int i=0;i<100;i++)
```
Can we correct this code to follow the style guidelines?

favre49 commented on doc/tutorials/reinforcement_learning/reinforcement_learning.txt:
```diff
@@ -231,9 +233,172 @@ to have converged when the average return reaches a predetermined value (i.e. >
+}
+@endcode
+
+What is "measure" here? It can be a lambda function which returns a boolean value(indicating the end of training)
```
```suggestion
What is "measure" here? It is a lambda function which returns a boolean value (indicating the end of training)
```

favre49 commented on doc/tutorials/reinforcement_learning/reinforcement_learning.txt:
```diff
@@ -231,9 +233,172 @@ to have converged when the average return reaches a predetermined value (i.e. >
+@endcode
+
+What is "measure" here? It can be a lambda function which returns a boolean value(indicating the end of training)
+and accepts the episode return(total reward of a deterministic test episode) as parameter.
```
```suggestion
and accepts the episode return (total reward of a deterministic test episode) as parameter.
```

favre49 commented on doc/tutorials/reinforcement_learning/reinforcement_learning.txt:
```diff
@@ -231,9 +233,172 @@ to have converged when the average return reaches a predetermined value (i.e. >
+
+What is "measure" here? It can be a lambda function which returns a boolean value(indicating the end of training)
+and accepts the episode return(total reward of a deterministic test episode) as parameter.
+So, lets create that.
```
```suggestion
So, let's create that.
```

favre49 commented on doc/tutorials/reinforcement_learning/reinforcement_learning.txt:
```diff
@@ -231,9 +233,172 @@ to have converged when the average return reaches a predetermined value (i.e. >
+  position = position % returns.n_elem;
+  episode++;
+
+  std::cout << "Episode No.: "        << episode 
```
```suggestion
  std::cout << "Episode No.: " << episode 
```

favre49 commented on doc/tutorials/reinforcement_learning/reinforcement_learning.txt:
```diff
@@ -231,9 +233,172 @@ to have converged when the average return reaches a predetermined value (i.e. >
+}
+@endcode
+
+It will train for 100 episodes, which will take around 50 seconds.
```
I think we can put this in the area where we explain the for loop?

favre49 commented on doc/tutorials/reinforcement_learning/reinforcement_learning.txt:
```diff
@@ -231,9 +233,172 @@ to have converged when the average return reaches a predetermined value (i.e. >
+Here is the full code, to try this right away:
+
+@code
+#include <mlpack/core.hpp>
```
Please correct this to conform to the style guidelines

joeljosephjin commented:
> If you're including sample code, can you add a test case matching it?
@birm Is it necessary? There is already a test case for Async Learning using One-Step Q-Learning Agents in "async_learning_test.cpp". And my sample code is almost same as that(just simplified some style of code to make understanding easier). Wouldn't adding another test case make the test file more clogged?

joeljosephjin commented:
> I'll see if anyone agrees with me, since I don't feel too strongly about it, but I do think that having a test would case would force us to keep the tutorial up to date.
> That said, maybe this isn't a tutorial, maybe it's an https://github.com/mlpack/examples ?

I have removed the sample code from here and replaced it with a link directing to mlpack/examples. This makes the tutorial file more clean, i hope. I'll do another pull request to mlpack/examples for the sample code. :)

That said, rest of the tutorial still uses some code snippets for explanation purposes.

thanks for the comments :)

birm commented:
Ah, I think I've given you bad advice here. I didn't want the code example to be removed, I'm just a little worried about it becoming wrong in the future. To that end, I thought either a small test, or an "asynchronous learning example" to go along with this would be a good idea. But maybe that's something for later. I'll approve if you revert b7108a3. Apologies for the confusion.

favre49 commented on doc/tutorials/reinforcement_learning/reinforcement_learning.txt:
```diff
@@ -231,9 +233,167 @@ to have converged when the average return reaches a predetermined value (i.e. >
+    position = position % returns.n_elem;
+    episode++;
+
+    std::cout << "Episode No.: "        << episode 
```
```suggestion
    std::cout << "Episode No.: " << episode 
```

favre49 commented:
I think we can go ahead and merge this then.
